### PR TITLE
Doc manager update

### PR DESCRIPTION
**Pull request by JD-Howard**
**Objective**
This is an update to the DocumentManager to persist/recycle more data. It was decided during the DefinitionProvider PR #45 that something had to be done about about the excessively expensive `ReadonlyDocument.atomsForest` being regenerated more often than it needed to be.

**Abstraction/Implementation**
This change involves a major emphasis on tracking pairs ReadonlyDocument's & vscode.TextDocuments and performing very cheap checks to detect when the ReadonlyDocument is out of date. It uses a set of flags to specify the various contextual applications. This will inevitably make a ReadonlyDocument qualify for up to 4 queries
- Currently Active Document
- Opened
- Project
- Workspace

However, the expensive parts should now be completely recycled and very cheap checks are in place to avoid the excessive parts of that **potential** duplication. From what I can tell, nothing should ever go through all 4 in expensive ways because any one would have passed the cheap pre-check, had the expensive work done and never moved on to the others.

The original version of this did a lot better job of purging dead data, but I believe this will at least make it inaccessible. This is notable for file renames and deletions performed directly in a OS file explorer. In theory, this is removing the flags that qualify them for various queries, but not necessarily removing the reference entirely across all possible situations. I believe the actions performed in the vscode workspace itself would do proper data purging purging though.

This also removed the PRJ embedded ReadonlyDocument property. Everything is demand queried from the DocumentManager when applicable, but persisted so it can efficiently serve it back next time. The cheap pre-check added in the DefinitionProvider and the fact that ReadonlyDocument creation is also quite cheap should adequately compensate for this. Note that PRJ's being opened/closed is probably the biggest issue with persisted dead data. Since the the ProjectTree's `Instance()` is always used to produce the "_single source of truth_" of fspath keys this should not actually produce erroneous PRJ references, but it has no mechanism for getting rid of the old ones. I'd like to maintain a maximum separation between the Document Manager and the Project Manager if we can.

Removed the attempts to persist data across workspace changes. From what I can tell, this was an exercise of futility. I tried multiple tests to verify it was working and kept eventually winding up with an empty `_cached` property. For example the current arrangement of the `initialize()` method that setups the cache isn't even clearing it anymore and all references are still disappearing after any action that changes the workspace.

Honestly the _documents.ts_ is vastly more clustered than it used to be since it is now handling various edge cases and was massively changed. So, I may be easier to perform your first review without the Adds/Deletes turned on.

**Tests Passed**
I kind of just described them above by identifying the potential known issues. I did run this with the F12 on a Workspace only, a PRJ only and a Workspace/PRJ hybrid.

**Screenshots**
Not Applicable

### CODE REVIEW

File: extension/src/documents.ts [hunk 1]
```diff
@@ -5,9 +5,37 @@ import { AutoLispExt } from './extension';
...
 
+enum Origins {
+	OPENED = 'O',
+	WSPACE = 'W',
+	PROJECT = 'P'
+}
+
+
+interface DocumentSources {
+	native: vscode.TextDocument;
```
Comment (hualin-adsk) [inline extension/src/documents.ts:16]: we need to give its more descriptive name, I think `rawDoc` or any others

File: extension/src/documents.ts [hunk 2]
```diff
@@ -5,9 +5,37 @@ import { AutoLispExt } from './extension';
...
+	internal: ReadonlyDocument;
+	flags: Set<Origins>;
```
Comment (hualin-adsk) [inline extension/src/documents.ts:18]: It means that for one document, it can be opened, in workspace and in project files? maybe we need to add a comment here

File: extension/src/documents.ts [hunk 3]
```diff
@@ -20,16 +48,63 @@ export class DocumentManager{
...
+		}
+		return result;
+	}
+
+	getDocument(nDoc: vscode.TextDocument) {
+		const key = nDoc.fileName.replace(/\//g, '\\');
+		if (this._cached.has(key)) {
+			const sources = this._cached.get(key);
+			sources.native = nDoc;
+			if (!sources.internal || !sources.internal.equal(sources.native)){
```
Comment (hualin-adsk) [inline extension/src/documents.ts:100]: I think this is the change we check if the atoms forest should regenerate

File: extension/src/documents.ts [hunk 4]
```diff
@@ -50,46 +125,66 @@ export class DocumentManager{
...
-			return ProjectTreeProvider.instance().projectNode.sourceFiles.map(x => 
-				this._opened.has(x.document.fileName) ? ReadonlyDocument.getMemoryDocument(this._opened.get(x.document.fileName)) : x.document
-				);
-		} else {
-			return [];
+			this.projectKeys.forEach(key => {
+				if (this._cached.has(key)){
+					const sources = this._cached.get(key);
+					sources.flags.add(Origins.PROJECT);
+					if (sources.native && !sources.internal.equal(sources.native)){
```
Comment (hualin-adsk) [inline extension/src/documents.ts:136]: I think this 3 lines codes are duplicated several times, maybe extract to a function

File: extension/src/documents.ts [hunk 5]
```diff
@@ -261,39 +413,37 @@ export class DocumentManager{
 		this.initialize();
 
 		AutoLispExt.Subscriptions.push(vscode.workspace.onDidCloseTextDocument((e: vscode.TextDocument) => {
-			if (this._opened.has(e.uri.fsPath) === true) {
-				this._opened.delete(e.uri.fsPath);
-			}
+			const source = this._cached.get(e.fileName.replace(/\//g, '\\'));
```
Comment (hualin-adsk) [inline extension/src/documents.ts:416]: e.fileName.replace(/\//g, '\\') also calls several times, maybe extract to a util small function

**Comment by hualin-adsk** [comment]
For F12 on a `setq`, it looks like it will find all the variables have the same name:
image

Maybe this is not an issue for F12, but it is risk for future F2 (rename)

File: extension/src/documents.ts [hunk 6]
```diff
@@ -5,9 +5,37 @@ import { AutoLispExt } from './extension';
 import { ProjectTreeProvider  } from "./project/projectTree";
 
 
+enum Origins {
+	OPENED = 'O',
```
[the remark this answers is not in the document]
Comment (JD-Howard) [reply extension/src/documents.ts:9]: These can go back to numbered enums now. I only added the strings to visually see which enum was getting added and purged during my tests. IE, I don't think we need these to have meaningful representations in the long term because they aren't exported.

File: extension/src/documents.ts (hunk 1 above)
Comment (JD-Howard) [reply extension/src/documents.ts:16]: In the context of this not being exported and living inside of `DocumentSources` interface, I actually think native and internal are meaningful. Clearly just an F2 to fix it if your comment isn't assuming use-cases beyond _documents.ts_ though.

File: extension/src/documents.ts (hunk 2 above)
Comment (JD-Howard) [reply extension/src/documents.ts:18]: Well, think about life cycle and the goal to recycle everything. 
Opening a file adds the OPENED flag, which is the only source of truth for what is actively opened. That file still could have had WSPACE & PRJ flags.
If you delete a file using file explorer while it was opened in the editor, the WSPACE flag gets removed, but we still have the reference with the OPENED flag and we should because VSCode doesn't close the document in that scenario.
If you are in a workspace, you open a document from file explorer, you then decide you need to close it, but add it to a PRJ, then nothing is lost.
If you open a PRJ, use F12, close that PRJ open a different one and later come back to it or open a different PRJ that also references that file, then nothing is lost.

File: extension/src/documents.ts (hunk 3 above)
Comment (JD-Howard) [reply extension/src/documents.ts:100]: Yes the `!sources.internal.equal(sources.native)` is the **cheap** check to see if there are disparities between our (internal) versions and the vscode (native) version of the document.

File: extension/src/documents.ts (hunk 4 above)
Comment (JD-Howard) [reply extension/src/documents.ts:136]: I know there is a lot of stuff that looks the same but sometimes has subtle differences. You are right, I should try a lot harder to clean this up with some meaningful function names even if they aren't recycled very much.

**Comment by JD-Howard** [comment]
> For F12 on a `setq`, it looks like it will find all the variables have the same name:
It should only ever do this when it couldn't find anything in clear context. This is a result of a last ditch effort/fallback plan of locating all the first level setq's within the document root Sexpressions. 

Re: F2
When `IsFunction=False`, the F2 will have to do a much deeper mapping of a documents function calls and variable localization's to properly do work. This scenario will only happen on the local document. I am still mixed on whether I want F2 doing work across opened documents, PRJ's and workspace documents for function names.

File: extension/src/documents.ts (hunk 5 above)
Comment (JD-Howard) [reply extension/src/documents.ts:416]: I will change these to some variant of:
`this._cached.get(this.normalizePath(e.fileName))`

File: extension/src/documents.ts (hunk 4 above)
Comment (JD-Howard) [reply extension/src/documents.ts]: I made 2.5 functions that consume the vast majority of the complexity. Should be a lot clearer now. These primary handlers do some unnecessary things/checks for certain circumstance, but I don't believe this hurts anything.

File: extension/src/documents.ts [hunk 7]
```diff
@@ -99,33 +206,44 @@ export class DocumentManager{
...
+				key = this.pathConsumeOrValidate(key, Origins.PROJECT);
+				if (key !== '') {
+					shouldExist.push(key);
 				}
 			});
+		}
+
+
+		/////////////////////// The extension appears to reset after every workspace change and this probably isn't really doing anything ///////////////////////
+		this.cacheKeys.filter(k => !shouldExist.includes(k)).forEach(obsolete => {
```
Comment (hualin-adsk) [inline extension/src/documents.ts:238]: we also noticed this behavior, the extension will reset if open a workspace, open a folder.

When reset the extension, I think the `cacheKeys` will get lost (or reset) automatically?

File: extension/src/providers/gotoProvider.ts [hunk 8]
```diff
@@ -8,7 +8,7 @@ import { SearchPatterns, SearchHandlers } from './providerShared';
 
 export class AutolispDefinitionProvider implements vscode.DefinitionProvider{
 	async provideDefinition(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): Promise<vscode.Location | vscode.Location[]> {
-		const rDoc = ReadonlyDocument.getMemoryDocument(document);
+		const rDoc = AutoLispExt.Documents.getDocument(document);
 		let selected = '';
```
Comment (hualin-adsk) [inline extension/src/providers/gotoProvider.ts:12]: For the line 32, I still want to filter out to re-search in the `ActiveDocument`

**Comment by hualin-adsk** [comment]
Thank you for refining this! 
I tried it in some real world cases:
 1. create project for all the `Express `tool in AutoCAD, F12 works well
 2. Open the lisp codes folder for AutoCAD Electrical which is written by lisp, it can work although the speed is slow for some large files.
I will try to use the `Chrome Dev tool `to see which part codes is the bottle neck codes.

File: extension/src/providers/gotoProvider.ts (hunk 8 above)
Comment (JD-Howard) [reply extension/src/providers/gotoProvider.ts:12]: I think we'll ultimately wind up as much and more likely a lot more work filter that out than it currently would failing the simple pre-check.

**Comment by JD-Howard** [comment]
I downloaded a bunch of random AutoLisp repo's and made a PRJ of the largest one. Think its largest file is just under 5000 lines. When I use F12 I get an infinite loop that never actually errors and once it gets going breakpoints don't function so it is really hard to find what is causing the problem. Who knows how exactly their code is randomly broken though.

**Comment by hualin-adsk** [comment]
Yes, maybe the parser is the bottle neck. I will do some profiling, it is not this PR work. :)

**Comment by hualin-adsk** [comment]
```
  (defun mode_tile_4_grp ( / )
    (cond
      ((= do_grp "1")
        (mode_tile "col_grp" 0)
      )
      (T (mode_tile "col_grp" 1))
    )
    (chk_accept_mode)
  )
```
If I select `mode_tile_4_grp` and hit F12, it will say there is no definition. I think this is a edge case.

**Comment by hualin-adsk** [comment]
And we can associate this PR to this issue https://github.com/Autodesk-AutoCAD/AutoLispExt/issues/30

File: extension/src/providers/gotoProvider.ts (hunk 8 above)
Comment (hualin-adsk) [reply extension/src/providers/gotoProvider.ts:12]: got it!

**Comment by JD-Howard** [comment]
> And for the variables definition, maybe it only takes care the local variables (defined in function, or function parameter); don't search global variables (it means that for variable to search in the CURRENT file is enough).

We can do this for now until the F2 functionality is available. I fully expect the contextual mapping of a document to get pretty drastic in order to have any real confidence in rename operations.

>If I select mode_tile_4_grp and hit F12, it will say there is no definition. I think this is an edge case.

I think i need slightly more context. I pasted this function into a blank document, added `(mode_tile_4_grp)` after it and it was located just fine. Can you email a snippet of this that doesn't work please.

File: extension/src/documents.ts (hunk 7 above)
Comment (JD-Howard) [reply extension/src/documents.ts:238]: Yes, the only reason I left it was because this actually does do all the work to reset the state. However, when it is done, it also appears that it completely resets the extension wasting that work. If you don't have any ideas on how to persist that data past the reset, then it can go away.

File: extension/src/documents.ts (hunk 1 above)
Comment (JD-Howard) [reply extension/src/documents.ts:17]: So, did you come around on this or are we going with roDoc & vsDoc?

Comment (hualin-adsk) [reply extension/src/documents.ts:17]: I am ok with the current name

File: extension/src/documents.ts (hunk 2 above)
Comment (hualin-adsk) [reply extension/src/documents.ts:19]: Thanks for explaining it!

File: extension/src/documents.ts (hunk 7 above)
Comment (hualin-adsk) [reply extension/src/documents.ts:238]: I don't have any ideal to persist the cache between the different folders. If the `cacheKeys` will be also reset automatically, I prefer to make the codes more simpler and cleaner (remove the `shouldExist`)

**Comment by hualin-adsk** [comment]
`>> I pasted this function into a blank document, added (mode_tile_4_grp) after it and it was located just fine`
I mean that to select the function name after the `defun`, then hit F12, it can repro the issue.
Don't add the callings like `(mode_tile_4_grp) `

**Comment by JD-Howard** [comment]
Your GIF illustrated an edge case problem in one of the new Sexpression functions. I fixed that and think everything else discussed was addressed. However, I am getting that infinite busy indicator on stuff I didn't previously have that problem with. So, I need to do some more investigating, but its certainly not going to be an easy thing to find without an actual error.

File: extension/src/providers/gotoProvider.ts [hunk 9]
```diff
@@ -84,9 +84,9 @@ export class AutolispDefinitionProvider implements vscode.DefinitionProvider{
 			}
 		} while (flag);
 		// If we still haven't found anything check the 1st occurrence of setq's. This will find globals setqs and nested ones possibly inside other defuns
-		if (result.length === 0){
```
Comment (hualin-adsk) [inline extension/src/providers/gotoProvider.ts:87]: try this codes, and it will not break in the above loop:
```
(defun test (num) 
  (print num)
  (print "I am in the same line")
  (print "hello")
  (inner (* 20 5))
)
(defun inner (var) 
  (print var) ;; try F12 on this var
  (print "world")
)
(setq str (test 20)) 
```

**Comment by hualin-adsk** [comment]
Try this simple case:
put this codes to a file and open that file, and it will not break in the above loop:

(defun test (num) 
  (print num)
  (print "I am in the same line")
  (print "hello")
  (inner (* 20 5))
)
(defun inner (var) 
  (print var) ;; try F12 on this var
  (print "world")
)
(setq str (test 20))

**Comment by hualin-adsk** [comment]
I think it is introduced by this change: https://github.com/Autodesk-AutoCAD/AutoLispExt/pull/49/commits/35afb0b8f1fde3608b70f1ba121d2b0ed99719fb#diff-604286706255a4a546d80ee16fd0a5263fae6e87f2a57f015b27870301f4858aL907

File: extension/src/providers/gotoProvider.ts (hunk 9 above)
Comment (JD-Howard) [reply extension/src/providers/gotoProvider.ts:87]: Yeah, I am currently researching that. I am thinking I shouldn't have fixed the Sexpression, I should have checked the variable for null and acted accordingly.

Comment (hualin-adsk) [reply extension/src/providers/gotoProvider.ts:87]: It looks like the changes in Sexpression is to fix the edge case for (defun test ()), not for the variables

**Comment by JD-Howard** [comment]
Global variables are working again, not scanning the whole document for Setq's and should now only be operating within the closest Defun of the the triggered position. Variables that are localized would never find a setq, they will always go up to the localization header. Does that sound like the behavior we want at this point? I expect this to change when the deep analysis for F2 comes in. 

IE, I would like to eventually make F12 to walk its way up to the Defun localization header; maybe even into another Defun that calls that function and also localizes that variable. Which I think will be easily possible after the F2 work is complete.

**Comment by hualin-adsk** [comment]
It looks like working well. And (defun) can contain another (defun), I will try to test this point.

**Comment by hualin-adsk** [comment]
Another issue:
```
 (setq lst (acet-geom-m-trans (list p1 p5 p6 p7 p1)
                     0 ;rk was na 6:28 PM 8/12/97
                     1
           );acet-geom-m-trans
 );setq
```
Select the comments `acet-geom-m-trans` and F12, it will report it can not find definition for `trans`

File: extension/src/providers/gotoProvider.ts [hunk 10]
```diff
@@ -77,32 +77,36 @@ export class AutolispDefinitionProvider implements vscode.DefinitionProvider{
...
 			} else {
 				context = parent;
 			}
 		} while (flag);
 		// If we still haven't found anything check the 1st occurrence of setq's. This will find globals setqs and nested ones possibly inside other defuns
-		// if (result.length === 0){
-		// 	this.findInSetqs(doc, ucName).forEach(x => { result.push(x); });
-		// }
+		if (result.length === 0){
+			const possible = searchIn.findChildren(SearchPatterns.DEFINES, true).filter(p => p.contains(start));
```
Comment (hualin-adsk) [inline extension/src/providers/gotoProvider.ts:88]: I think this is the key codes to narrow the setq search

File: extension/src/providers/gotoProvider.ts [hunk 11]
```diff
@@ -77,32 +77,36 @@ export class AutolispDefinitionProvider implements vscode.DefinitionProvider{
...
 
-	private findInSetqs(doc: ReadonlyDocument, ucName: string): vscode.Location[] {
+	private findInSetqs(sexp: Sexpression, ucName: string, fileName: string): vscode.Location[] {
 		const result: vscode.Location[] = [];
-		doc.findExpressions(SearchPatterns.ASSIGNS).forEach(setq => {
+		const found = sexp.findChildren(SearchPatterns.ASSIGNS, false);
+		found.forEach(setq => {
 			let isVar = false;
-			let cIndex = setq.nextKeyIndex(0);
+			let cIndex = setq.nextKeyIndex(0, true);
```
Comment (hualin-adsk) [inline extension/src/providers/gotoProvider.ts:101]: bool parameter hurts the codes readability.

**Review by hualin-adsk — APPROVED** [review]
